### ISSUE TITLE
Restricting some actions from accessing and triggering via transport layer

### ISSUE THREAD

**Issue by dkulichkin** [senecajs/seneca#122]
Is there any way to maintain it currently or plans to implement soon?

Thanks

[Trello Card](https://trello.com/c/W9SCOk3F/94-seneca-restricting-some-actions-from-accessing-and-triggering-via-transport-layer)

**Comment by mcdonnelldean** [comment]
Hi @dkulichkin,

Not on any of the transports specifically, you could wrap one yourself. There is the [seneca-perm](https://github.com/rjrodger/seneca-perm) that limits users access to actions. Not exactly what you are looking for but it would be a good basis to start.

Could you give me some detail on your exact requirements. If nothing someone else might take it up as a plugin to build, you never know.

**Comment by dkulichkin** [comment]
In my particular case I've part of a business logic implemented as a plugin with actions supporting besides the regular CRUD some servicing operations such ones like locking/unlocking the jobs, sending notifications, etc. With that the former ones could only be triggered internally by the code's execution workflow so I don't want to expose them as publicly accessible. There's no also any particular permission role to delegate it to though - as I said it's only the execution workflow's conditions.

**Comment by mcdonnelldean** [comment]
If it was me, I wouldn't go as far as making a new plugin (although you're encouraged to do so). I'd use `seneca-perm` with an admin or internal user and handle it that way. The nice thing about this is it would allow you to create fine grained permissions later on.

You don't even need to use a DB. You could hard code the creation of a single user with permissions to be saved in memory. Given it is only one user there isn't much point going further with storage.

So basically:

- make hardcoded user (myservice-admin)
- use seneca-perm to limit actions to this user
- profit

Where I find this to be the best case is usually down the line you need to ability to work with restricted API's usually from a demo or admin point of view. Having a system that supports getting to these options is more preferable in the long run than simply not allowing them to ever be called.

I have a couple of services built where I have certain actions fired off to trigger analysis and things like that, usually when a condition is met. I don't want consumers ever being able to run these but when I built a little web tool to test the service under load and do general admin, it turned out nice to be able to force these actions to be ran by authenticating as a user in the same set as my admin-user for the service.

**Comment by dkulichkin** [comment]
Thanks Dean, I'll familiarize myself with seneca-perm

**Comment by rjrodger** [comment]
Transports need to be more easily composable.

Simple way to do this might be to add config option to transports, say: register:false, that prevents registration of the transport pin in the clientrouter of the seneca instance. This means that a child transport instantiated from a parent, would not interfere with client routing.

Would make this easier: https://github.com/mmalecki/seneca-loadbalance-transport

**Comment by r4j4h** [comment]
@mcdonnelldean Thank you, very cool idea.

@rjrodger Should a new Issue be made to represent this?

**Comment by mcdonnelldean** [comment]
Merged with #399 in terms of transport composability. It should in theory be much easier to do the above once it lands.
